File: Week 03/id_262/valid-perfect-square.py
class Solution:
    def isPerfectSquare(self, num: int) -> bool:
        if num == 0 or num == 1:
            return True
        start = 0
        end = num
        b = None

        while start < end:
            mid = (end + start) >> 1

            if mid == 0 or mid == b:
                return False

            if mid * mid > num:
                end = mid
            elif mid * mid < num:
                start = mid
            elif mid * mid == num:
                return True
            b = mid

File: Week 03/id_362/ValidPerfectSquare.swift
class Solution {
    func isPerfectSquare(_ num: Int) -> Bool {
        var left = 0
        var right = num / 2 + 1
        while left < right {
            let mid = (left + right + 1)  / 2
            let square = mid * mid
            if square > num {
                right = mid - 1
            } else {
                left = mid
            }
        }

        return left * left == num
    }
}

assert(Solution().isPerfectSquare(14) == false)
assert(Solution().isPerfectSquare(4) == true)

